client/lib/streamframer/framer.go

```diff
@@ -55,30 +55,45 @@ func (s *StreamFrame) IsCleared() bool {
 	}
 }
 
+func (s *StreamFrame) Copy() *StreamFrame {
+	n := new(StreamFrame)
+	*n = *s
+	n.Data = make([]byte, len(s.Data))
+	copy(n.Data, s.Data)
```

Comment:
Given the race on the backing slice I understand how this fixes the issue but it is rather unfortunate that we are copying the data! I believe another fix would be for the out channel of the framer to be a wrapper:

```
type FrameFuture struct {
  Frame *StreamFrame
  FrameSent chan struct{}
}

func (f *FrameFuture) Sent() {
  close(f.FrameSent)
}
```

And then the caller can call Sent on the future after it has been flushed: https://github.com/hashicorp/nomad/pull/4234/files#diff-03a99c783d6606b25fccfc5f3d26092bR470

Then send() can select on that and the exit channel. This way we can avoid copying.